File: domain/src/main/java/com/arturogutierrez/openticator/domain/otp/TimeBasedOneTimePassword.java
package com.arturogutierrez.openticator.domain.otp;

import com.arturogutierrez.openticator.domain.otp.model.Passcode;
import com.arturogutierrez.openticator.domain.otp.time.TimeCalculator;

class TimeBasedOneTimePassword implements OneTimePassword {

  private final OneTimePasswordGenerator oneTimePasswordGenerator;
  private final TimeCalculator timeCalculator;

  public TimeBasedOneTimePassword(OneTimePasswordGenerator oneTimePasswordGenerator,
      TimeCalculator timeCalculator) {
    this.oneTimePasswordGenerator = oneTimePasswordGenerator;
    this.timeCalculator = timeCalculator;
  }

  @Override
  public Passcode generate() {
    long currentStep = timeCalculator.getCurrentTimeStep();
    long validUntilInSeconds = timeCalculator.getValidUntilInSeconds(currentStep);
    String code = oneTimePasswordGenerator.generate(currentStep);
    return new Passcode(code, validUntilInSeconds);
  }
}

File: domain/src/main/kotlin/com/arturogutierrez/openticator/domain/otp/TimeBasedOneTimePassword.kt
package com.arturogutierrez.openticator.domain.otp

import com.arturogutierrez.openticator.domain.otp.model.Passcode
import com.arturogutierrez.openticator.domain.otp.time.TimeCalculator
import com.arturogutierrez.openticator.domain.otp.time.TimeProvider

internal class TimeBasedOneTimePassword(private val currentTimeProvider: TimeProvider,
                                        private val oneTimePasswordGenerator: OneTimePasswordGenerator,
                                        private val timeCalculator: TimeCalculator) : OneTimePassword {

    override fun generate(): Passcode {
        val currentTimeInSeconds = currentTimeProvider.currentTimeInSeconds()
        val currentStep = timeCalculator.getCurrentTimeStep(currentTimeInSeconds)
        val validUntilInSeconds = timeCalculator.getValidUntilInSeconds(currentStep)
        val code = oneTimePasswordGenerator.generate(currentStep)
        return Passcode(code, validUntilInSeconds)
    }
}
